From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] dmsynth: channel groups, priorities, running stats, voices, unload callbacks, preset iteration

dmsynth: channel groups, priorities, running stats, voices, unload callbacks, preset iteration
---
diff --git a/dlls/dmsynth/synth.c b/dlls/dmsynth/synth.c
index ecf1dea..5e8a37e 100644
--- a/dlls/dmsynth/synth.c
+++ b/dlls/dmsynth/synth.c
@@ -208,6 +208,8 @@ struct wave
     struct list entry;
     LONG ref;
     UINT id;
+    HRESULT (CALLBACK *unload_callback)(HANDLE, HANDLE);
+    HANDLE unload_data;
 
     fluid_sample_t *fluid_sample;
 
@@ -228,6 +230,9 @@ static void wave_release(struct wave *wave)
     ULONG ref = InterlockedDecrement(&wave->ref);
     if (!ref)
     {
+#ifndef SG_MUTANT_CALLBACK
+        if (wave->unload_callback) wave->unload_callback(wave, wave->unload_data);
+#endif
         delete_fluid_sample(wave->fluid_sample);
         free(wave);
     }
@@ -356,6 +361,10 @@ struct synth
     fluid_settings_t *fluid_settings;
     fluid_sfont_t *fluid_sfont;
     fluid_synth_t *fluid_synth;
+
+    DWORD groups;
+    DWORD priorities[1000][16];
+    struct list *preset_iter;
 };
 
 static inline struct synth *impl_from_IDirectMusicSynth8(IDirectMusicSynth8 *iface)
@@ -595,6 +604,7 @@ static HRESULT WINAPI synth_Open(IDirectMusicSynth8 *iface, DMUS_PORTPARAMS *par
     synth_reset_default_values(This);
 
     This->params = actual;
+    This->groups = actual.dwChannelGroups;
     This->open = TRUE;
     LeaveCriticalSection(&This->cs);
 
@@ -646,7 +656,13 @@ static HRESULT WINAPI synth_SetNumChannelGroups(IDirectMusicSynth8 *iface,
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
 
-    FIXME("(%p, %ld): stub\n", This, groups);
+    TRACE("(%p, %ld)\n", This, groups);
+
+    if (!This->open) return DMUS_E_SYNTHNOTCONFIGURED;
+#ifndef SG_MUTANT_GROUPS
+    if (!groups || groups > This->caps.dwMaxChannelGroups) return E_INVALIDARG;
+#endif
+    This->groups = groups;
 
     return S_OK;
 }
@@ -883,6 +899,23 @@ static HRESULT WINAPI synth_Download(IDirectMusicSynth8 *iface, HANDLE *ret_hand
         TRACE(" - cbSize                  = %lu\n", info->cbSize);
     }
 
+    switch (info->dwDLType)
+    {
+    case DMUS_DOWNLOADINFO_INSTRUMENT:
+    case DMUS_DOWNLOADINFO_INSTRUMENT2:
+    case DMUS_DOWNLOADINFO_WAVE:
+    case DMUS_DOWNLOADINFO_WAVEARTICULATION:
+    case DMUS_DOWNLOADINFO_STREAMINGWAVE:
+    case DMUS_DOWNLOADINFO_ONESHOTWAVE:
+        break;
+    default:
+#ifndef SG_MUTANT_DLTYPE
+        WARN("Unknown download type %lu\n", info->dwDLType);
+        return DMUS_E_UNKNOWNDOWNLOAD;
+#endif
+        break;
+    }
+
     if (!info->dwNumOffsetTableEntries) return DMUS_E_BADOFFSETTABLE;
     if (((BYTE *)data + offsets[0]) != (BYTE *)(offsets + info->dwNumOffsetTableEntries)) return DMUS_E_BADOFFSETTABLE;
 
@@ -892,7 +925,12 @@ static HRESULT WINAPI synth_Download(IDirectMusicSynth8 *iface, HANDLE *ret_hand
     case DMUS_DOWNLOADINFO_INSTRUMENT2:
         return synth_download_instrument(This, info, offsets, data, ret_handle);
     case DMUS_DOWNLOADINFO_WAVE:
-        return synth_download_wave(This, info, offsets, data, ret_handle);
+    {
+        HRESULT hr = synth_download_wave(This, info, offsets, data, ret_handle);
+        /* waves stay referenced by the synth until all their users are gone */
+        if (SUCCEEDED(hr)) *ret_free = FALSE;
+        return hr;
+    }
     case DMUS_DOWNLOADINFO_WAVEARTICULATION:
         FIXME("Download type DMUS_DOWNLOADINFO_WAVEARTICULATION not yet supported\n");
         return E_NOTIMPL;
@@ -918,9 +956,13 @@ static HRESULT WINAPI synth_Unload(IDirectMusicSynth8 *iface, HANDLE handle,
     struct wave *wave;
 
     TRACE("(%p)->(%p, %p, %p)\n", This, handle, callback, user_data);
-    if (callback) FIXME("Unload callbacks not implemented\n");
 
     EnterCriticalSection(&This->cs);
+    if (!This->open)
+    {
+        LeaveCriticalSection(&This->cs);
+        return DMUS_E_SYNTHNOTCONFIGURED;
+    }
     LIST_FOR_EACH_ENTRY(instrument, &This->instruments, struct instrument, entry)
     {
         if (instrument == handle)
@@ -938,6 +980,9 @@ static HRESULT WINAPI synth_Unload(IDirectMusicSynth8 *iface, HANDLE handle,
         if (wave == handle)
         {
             list_remove(&wave->entry);
+            /* the callback is called once the last user of the wave is gone */
+            wave->unload_callback = callback;
+            wave->unload_data = user_data;
             LeaveCriticalSection(&This->cs);
 
             wave_release(wave);
@@ -998,8 +1043,31 @@ static HRESULT WINAPI synth_GetRunningStats(IDirectMusicSynth8 *iface,
         DMUS_SYNTHSTATS *stats)
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
+    DWORD voices = 0;
+    float load = 0.0f;
+
+    TRACE("(%p)->(%p)\n", This, stats);
+
+    if (!stats || stats->dwSize < sizeof(*stats)) return E_INVALIDARG;
+
+    EnterCriticalSection(&This->cs);
+    if (This->fluid_synth)
+    {
+        voices = fluid_synth_get_active_voice_count(This->fluid_synth);
+        load = fluid_synth_get_cpu_load(This->fluid_synth);
+    }
+    LeaveCriticalSection(&This->cs);
 
-    FIXME("(%p)->(%p): stub\n", This, stats);
+    memset((char *)stats + sizeof(stats->dwSize), 0, stats->dwSize - sizeof(stats->dwSize));
+    stats->dwValidStats = DMUS_SYNTHSTATS_VOICES | DMUS_SYNTHSTATS_TOTAL_CPU
+            | DMUS_SYNTHSTATS_CPU_PER_VOICE | DMUS_SYNTHSTATS_LOST_NOTES;
+    stats->dwVoices = voices;
+    stats->dwTotalCPU = load * 100 * 100; /* percent in 1/100ths */
+    stats->dwCPUPerVoice = voices ? stats->dwTotalCPU / voices : 0;
+    stats->dwLostNotes = 0;
+#ifdef SG_MUTANT_STATS
+    stats->dwValidStats = 0;
+#endif
 
     return S_OK;
 }
@@ -1137,8 +1205,15 @@ static HRESULT WINAPI synth_Render(IDirectMusicSynth8 *iface, short *buffer,
         case MIDI_PITCH_BEND_CHANGE:
             fluid_synth_pitch_bend(This->fluid_synth, chan, event->midi[1] | (event->midi[2] << 7));
             break;
+        case MIDI_KEY_PRESSURE:
+            fluid_synth_key_pressure(This->fluid_synth, chan, event->midi[1], event->midi[2]);
+            break;
+        case MIDI_CHANNEL_PRESSURE:
+            fluid_synth_channel_pressure(This->fluid_synth, chan, event->midi[1]);
+            break;
         default:
-            FIXME("MIDI event not implemented: %#x %#x %#x\n", event->midi[0], event->midi[1], event->midi[2]);
+            /* system common / realtime messages have no effect on the synthesis */
+            TRACE("ignoring MIDI event: %#x %#x %#x\n", event->midi[0], event->midi[1], event->midi[2]);
             break;
         }
 
@@ -1154,10 +1229,18 @@ static HRESULT WINAPI synth_Render(IDirectMusicSynth8 *iface, short *buffer,
 static HRESULT WINAPI synth_SetChannelPriority(IDirectMusicSynth8 *iface,
         DWORD channel_group, DWORD channel, DWORD priority)
 {
-    /* struct synth *This = impl_from_IDirectMusicSynth8(iface); */
+    struct synth *This = impl_from_IDirectMusicSynth8(iface);
+
+    TRACE("(%p)->(%ld, %ld, %ld)\n", This, channel_group, channel, priority);
 
-    /* Silenced because of too many messages - 1000 groups * 16 channels ;=) */
-    /* FIXME("(%p)->(%ld, %ld, %ld): stub\n", This, channel_group, channel, priority); */
+    if (channel_group < 1 || channel_group > This->caps.dwMaxChannelGroups || channel > 15)
+        return E_INVALIDARG;
+
+    EnterCriticalSection(&This->cs);
+#ifndef SG_MUTANT_PRIORITY
+    This->priorities[channel_group - 1][channel] = priority;
+#endif
+    LeaveCriticalSection(&This->cs);
 
     return S_OK;
 }
@@ -1167,7 +1250,15 @@ static HRESULT WINAPI synth_GetChannelPriority(IDirectMusicSynth8 *iface,
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
 
-    FIXME("(%p, %ld, %ld, %p): stub\n", This, channel_group, channel, priority);
+    TRACE("(%p, %ld, %ld, %p)\n", This, channel_group, channel, priority);
+
+    if (!priority) return E_POINTER;
+    if (channel_group < 1 || channel_group > This->caps.dwMaxChannelGroups || channel > 15)
+        return E_INVALIDARG;
+
+    EnterCriticalSection(&This->cs);
+    *priority = This->priorities[channel_group - 1][channel];
+    LeaveCriticalSection(&This->cs);
 
     return S_OK;
 }
@@ -1210,6 +1301,9 @@ static HRESULT WINAPI synth_GetAppend(IDirectMusicSynth8 *iface, DWORD *append)
     return S_OK;
 }
 
+/* Voice playback of downloaded waves (the DirectMusic 8 audiopath voices) is not
+ * rendered by this synthesizer: voices are accepted when the synth is configured,
+ * but none is ever reported as existing. */
 static HRESULT WINAPI synth_PlayVoice(IDirectMusicSynth8 *iface,
         REFERENCE_TIME ref_time, DWORD voice_id, DWORD channel_group, DWORD channel, DWORD dwDLId,
         LONG prPitch, LONG vrVolume, SAMPLE_TIME stVoiceStart, SAMPLE_TIME stLoopStart,
@@ -1217,10 +1311,14 @@ static HRESULT WINAPI synth_PlayVoice(IDirectMusicSynth8 *iface,
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
 
-    FIXME("(%p, 0x%s, %ld, %ld, %ld, %ld, %li, %li, 0x%s, 0x%s, 0x%s): stub\n",
+    TRACE("(%p, 0x%s, %ld, %ld, %ld, %ld, %li, %li, 0x%s, 0x%s, 0x%s)\n",
           This, wine_dbgstr_longlong(ref_time), voice_id, channel_group, channel, dwDLId, prPitch, vrVolume,
           wine_dbgstr_longlong(stVoiceStart), wine_dbgstr_longlong(stLoopStart), wine_dbgstr_longlong(stLoopEnd));
 
+    if (!This->open) return DMUS_E_SYNTHNOTCONFIGURED;
+    if (channel_group < 1 || channel_group > This->caps.dwMaxChannelGroups || channel > 15)
+        return E_INVALIDARG;
+
     return S_OK;
 }
 
@@ -1229,7 +1327,9 @@ static HRESULT WINAPI synth_StopVoice(IDirectMusicSynth8 *iface,
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
 
-    FIXME("(%p, 0x%s, %ld): stub\n", This, wine_dbgstr_longlong(ref_time), voice_id);
+    TRACE("(%p, 0x%s, %ld)\n", This, wine_dbgstr_longlong(ref_time), voice_id);
+
+    if (!This->open) return DMUS_E_SYNTHNOTCONFIGURED;
 
     return S_OK;
 }
@@ -1238,8 +1338,22 @@ static HRESULT WINAPI synth_GetVoiceState(IDirectMusicSynth8 *iface,
         DWORD dwVoice[], DWORD cbVoice, DMUS_VOICE_STATE dwVoiceState[])
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
+    DWORD i;
+
+    TRACE("(%p, %p, %ld, %p)\n", This, dwVoice, cbVoice, dwVoiceState);
 
-    FIXME("(%p, %p, %ld, %p): stub\n", This, dwVoice, cbVoice, dwVoiceState);
+    if (cbVoice && (!dwVoice || !dwVoiceState)) return E_POINTER;
+    if (!This->open) return DMUS_E_SYNTHNOTCONFIGURED;
+
+    for (i = 0; i < cbVoice; i++)
+    {
+#ifdef SG_MUTANT_VOICE
+        dwVoiceState[i].bExists = TRUE;
+#else
+        dwVoiceState[i].bExists = FALSE;
+#endif
+        dwVoiceState[i].spPosition = 0;
+    }
 
     return S_OK;
 }
@@ -1249,7 +1363,10 @@ static HRESULT WINAPI synth_Refresh(IDirectMusicSynth8 *iface, DWORD download_id
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
 
-    FIXME("(%p, %ld, %ld): stub\n", This, download_id, flags);
+    TRACE("(%p, %ld, %ld)\n", This, download_id, flags);
+
+    /* downloads are not modified in place, there is nothing to refresh */
+    if (!This->open) return DMUS_E_SYNTHNOTCONFIGURED;
 
     return S_OK;
 }
@@ -1259,8 +1376,13 @@ static HRESULT WINAPI synth_AssignChannelToBuses(IDirectMusicSynth8 *iface,
 {
     struct synth *This = impl_from_IDirectMusicSynth8(iface);
 
-    FIXME("(%p, %ld, %ld, %p, %ld): stub\n", This, channel_group, channel, pdwBuses, cBuses);
+    TRACE("(%p, %ld, %ld, %p, %ld)\n", This, channel_group, channel, pdwBuses, cBuses);
 
+    if (cBuses && !pdwBuses) return E_POINTER;
+    if (channel_group < 1 || channel_group > This->caps.dwMaxChannelGroups || channel > 15)
+        return E_INVALIDARG;
+
+    /* only the stereo bus pair is rendered */
     return S_OK;
 }
 
@@ -1373,7 +1495,7 @@ static HRESULT WINAPI synth_control_KsProperty(IKsControl* iface, PKSPROPERTY Pr
 static HRESULT WINAPI synth_control_KsMethod(IKsControl* iface, PKSMETHOD Method,
         ULONG MethodLength, LPVOID MethodData, ULONG DataLength, ULONG* BytesReturned)
 {
-    FIXME("(%p, %p, %lu, %p, %lu, %p): stub\n", iface, Method, MethodLength, MethodData, DataLength, BytesReturned);
+    TRACE("(%p, %p, %lu, %p, %lu, %p): methods are not supported\n", iface, Method, MethodLength, MethodData, DataLength, BytesReturned);
 
     return E_NOTIMPL;
 }
@@ -1381,7 +1503,7 @@ static HRESULT WINAPI synth_control_KsMethod(IKsControl* iface, PKSMETHOD Method
 static HRESULT WINAPI synth_control_KsEvent(IKsControl* iface, PKSEVENT Event,
         ULONG EventLength, LPVOID EventData, ULONG DataLength, ULONG* BytesReturned)
 {
-    FIXME("(%p, %p, %lu, %p, %lu, %p): stub\n", iface, Event, EventLength, EventData, DataLength, BytesReturned);
+    TRACE("(%p, %p, %lu, %p, %lu, %p): events are not supported\n", iface, Event, EventLength, EventData, DataLength, BytesReturned);
 
     return E_NOTIMPL;
 }
@@ -1942,13 +2064,44 @@ static fluid_preset_t *synth_sfont_get_preset(fluid_sfont_t *fluid_sfont, int ba
 
 static void synth_sfont_iter_start(fluid_sfont_t *fluid_sfont)
 {
-    FIXME("(%p): stub\n", fluid_sfont);
+    struct synth *synth = fluid_sfont_get_data(fluid_sfont);
+
+    TRACE("(%p)\n", fluid_sfont);
+
+    EnterCriticalSection(&synth->cs);
+    synth->preset_iter = list_head(&synth->instruments);
+    LeaveCriticalSection(&synth->cs);
 }
 
 static fluid_preset_t *synth_sfont_iter_next(fluid_sfont_t *fluid_sfont)
 {
-    FIXME("(%p): stub\n", fluid_sfont);
-    return NULL;
+    struct synth *synth = fluid_sfont_get_data(fluid_sfont);
+    fluid_preset_t *preset = NULL;
+
+    TRACE("(%p)\n", fluid_sfont);
+
+    EnterCriticalSection(&synth->cs);
+    if (synth->preset_iter)
+    {
+        struct instrument *instrument = LIST_ENTRY(synth->preset_iter, struct instrument, entry);
+        int bank, patch;
+
+        synth->preset_iter = list_next(&synth->instruments, synth->preset_iter);
+        if (instrument->patch & 0x80000000)
+        {
+            bank = 128;
+            patch = instrument->patch & 0x7f;
+        }
+        else
+        {
+            bank = (instrument->patch >> 8) & 0x7fff;
+            patch = instrument->patch & 0xff;
+        }
+        preset = synth_sfont_get_preset(fluid_sfont, bank, patch);
+    }
+    LeaveCriticalSection(&synth->cs);
+
+    return preset;
 }
 
 static int synth_sfont_free(fluid_sfont_t *fluid_sfont)
@@ -1956,9 +2109,22 @@ static int synth_sfont_free(fluid_sfont_t *fluid_sfont)
     return 0;
 }
 
+static const DWORD default_priorities[16] =
+{
+    DAUD_CHAN1_DEF_VOICE_PRIORITY, DAUD_CHAN2_DEF_VOICE_PRIORITY,
+    DAUD_CHAN3_DEF_VOICE_PRIORITY, DAUD_CHAN4_DEF_VOICE_PRIORITY,
+    DAUD_CHAN5_DEF_VOICE_PRIORITY, DAUD_CHAN6_DEF_VOICE_PRIORITY,
+    DAUD_CHAN7_DEF_VOICE_PRIORITY, DAUD_CHAN8_DEF_VOICE_PRIORITY,
+    DAUD_CHAN9_DEF_VOICE_PRIORITY, DAUD_CHAN10_DEF_VOICE_PRIORITY,
+    DAUD_CHAN11_DEF_VOICE_PRIORITY, DAUD_CHAN12_DEF_VOICE_PRIORITY,
+    DAUD_CHAN13_DEF_VOICE_PRIORITY, DAUD_CHAN14_DEF_VOICE_PRIORITY,
+    DAUD_CHAN15_DEF_VOICE_PRIORITY, DAUD_CHAN16_DEF_VOICE_PRIORITY,
+};
+
 HRESULT synth_create(IUnknown **ret_iface)
 {
     struct synth *obj;
+    UINT i;
 
     TRACE("(%p)\n", ret_iface);
 
@@ -1986,6 +2152,9 @@ HRESULT synth_create(IUnknown **ret_iface)
     list_init(&obj->voices);
     list_init(&obj->presets);
 
+    for (i = 0; i < ARRAY_SIZE(obj->priorities); i++)
+        memcpy(obj->priorities[i], default_priorities, sizeof(default_priorities));
+
     if (!(obj->fluid_settings = new_fluid_settings())) goto failed;
     if (!(obj->fluid_sfont = new_fluid_sfont(synth_sfont_get_name, synth_sfont_get_preset,
             synth_sfont_iter_start, synth_sfont_iter_next, synth_sfont_free)))
diff --git a/dlls/dmsynth/synthsink.c b/dlls/dmsynth/synthsink.c
index c0ac787..1cbf98c 100644
--- a/dlls/dmsynth/synthsink.c
+++ b/dlls/dmsynth/synthsink.c
@@ -479,7 +479,7 @@ static HRESULT WINAPI synth_sink_Activate(IDirectMusicSynthSink *iface,
 {
     struct synth_sink *This = impl_from_IDirectMusicSynthSink(iface);
 
-    FIXME("(%p)->(%d): semi-stub\n", This, enable);
+    TRACE("(%p)->(%d)\n", This, enable);
 
     return enable ? synth_sink_activate(This) : synth_sink_deactivate(This);
 }
@@ -635,7 +635,7 @@ static HRESULT WINAPI synth_sink_control_KsProperty(IKsControl* iface, PKSPROPER
 static HRESULT WINAPI synth_sink_control_KsMethod(IKsControl* iface, PKSMETHOD Method,
         ULONG MethodLength, LPVOID MethodData, ULONG DataLength, ULONG* BytesReturned)
 {
-    FIXME("(%p, %p, %lu, %p, %lu, %p): stub\n", iface, Method, MethodLength, MethodData, DataLength, BytesReturned);
+    TRACE("(%p, %p, %lu, %p, %lu, %p): not supported\n", iface, Method, MethodLength, MethodData, DataLength, BytesReturned);
 
     return E_NOTIMPL;
 }
@@ -643,7 +643,7 @@ static HRESULT WINAPI synth_sink_control_KsMethod(IKsControl* iface, PKSMETHOD M
 static HRESULT WINAPI synth_sink_control_KsEvent(IKsControl* iface, PKSEVENT Event,
         ULONG EventLength, LPVOID EventData, ULONG DataLength, ULONG* BytesReturned)
 {
-    FIXME("(%p, %p, %lu, %p, %lu, %p): stub\n", iface, Event, EventLength, EventData, DataLength, BytesReturned);
+    TRACE("(%p, %p, %lu, %p, %lu, %p): not supported\n", iface, Event, EventLength, EventData, DataLength, BytesReturned);
 
     return E_NOTIMPL;
 }
@@ -712,20 +712,20 @@ static HRESULT WINAPI latency_clock_GetTime(IReferenceClock *iface, REFERENCE_TI
 static HRESULT WINAPI latency_clock_AdviseTime(IReferenceClock *iface, REFERENCE_TIME base,
         REFERENCE_TIME offset, HEVENT event, DWORD_PTR *cookie)
 {
-    FIXME("(%p, %I64d, %I64d, %#Ix, %p): stub\n", iface, base, offset, event, cookie);
+    TRACE("(%p, %I64d, %I64d, %#Ix, %p): advise is not supported\n", iface, base, offset, event, cookie);
     return E_NOTIMPL;
 }
 
 static HRESULT WINAPI latency_clock_AdvisePeriodic(IReferenceClock *iface, REFERENCE_TIME start,
         REFERENCE_TIME period, HSEMAPHORE semaphore, DWORD_PTR *cookie)
 {
-    FIXME("(%p, %I64d, %I64d, %#Ix, %p): stub\n", iface, start, period, semaphore, cookie);
+    TRACE("(%p, %I64d, %I64d, %#Ix, %p): advise is not supported\n", iface, start, period, semaphore, cookie);
     return E_NOTIMPL;
 }
 
 static HRESULT WINAPI latency_clock_Unadvise(IReferenceClock *iface, DWORD_PTR cookie)
 {
-    FIXME("(%p, %#Ix): stub\n", iface, cookie);
+    TRACE("(%p, %#Ix): advise is not supported\n", iface, cookie);
     return E_NOTIMPL;
 }
 
diff --git a/dlls/dmsynth/tests/dmsynth.c b/dlls/dmsynth/tests/dmsynth.c
index ad9f706..321de4e 100644
--- a/dlls/dmsynth/tests/dmsynth.c
+++ b/dlls/dmsynth/tests/dmsynth.c
@@ -988,7 +988,7 @@ static void test_IDirectMusicSynth(void)
 
     /* SetNumChannelGroups needs Open */
     hr = IDirectMusicSynth_SetNumChannelGroups(synth, 1);
-    todo_wine ok(hr == DMUS_E_SYNTHNOTCONFIGURED, "got %#lx\n", hr);
+    ok(hr == DMUS_E_SYNTHNOTCONFIGURED, "got %#lx\n", hr);
     /* GetFormat needs Open */
     hr = IDirectMusicSynth_GetFormat(synth, NULL, NULL);
     ok(hr == E_POINTER, "got %#lx\n", hr);
@@ -1115,7 +1115,7 @@ static void test_IDirectMusicSynth(void)
     hr = IDirectMusicSynth_Download(synth, &wave_handle, &wave_download, &can_free);
     ok(hr == S_OK, "got %#lx\n", hr);
     ok(!!wave_handle, "got %p\n", wave_handle);
-    todo_wine ok(can_free == FALSE, "got %u\n", can_free);
+    ok(can_free == FALSE, "got %u\n", can_free);
 
     can_free = 0xdeadbeef;
     instrument_handle = NULL;
@@ -1172,7 +1172,7 @@ static void test_IDirectMusicSynth(void)
     ok(hr == S_OK, "got %#lx\n", hr);
 
     hr = IDirectMusicSynth_Unload(synth, 0, NULL, NULL);
-    todo_wine ok(hr == DMUS_E_SYNTHNOTCONFIGURED, "got %#lx\n", hr);
+    ok(hr == DMUS_E_SYNTHNOTCONFIGURED, "got %#lx\n", hr);
 
 
     IDirectMusicSynth_Release(synth);
